# Part 1: Sampling & Quantization
For each signal we choose an ADC input range equal to the signal’s min→max, and a 10-bit resolution (1024 levels), which yields a quantization step Δ≈range/1000.

In [ ]:
import numpy as np
import math

# Parameters
files = ['signal1.csv', 'signal2.csv', 'signal3.csv']
data_dir = '../data/'       
target_levels = 1000        

print(f"{'Signal':<10}{'Vmin':>8}{'Vmax':>8}{'Range':>9}{'Δ (step)':>12}{'Bits':>8}")
print('-'*55)

for fname in files:
    # 1. Load data
    data = np.loadtxt(f"{data_dir}{fname}", delimiter=',')
    
    # 2. Compute min, max, range
    vmin, vmax = data.min(), data.max()
    rng = vmax - vmin
    
    # 3. Choose quantization step Δ
    delta = rng / target_levels
    
    # 4. Compute required bits N
    levels_needed = rng / delta
    N = math.ceil(math.log2(levels_needed + 1))
    
    # 5. Print 
    signal_label = fname.replace('.csv','')
    print(f"{signal_label:<10}{vmin:8.3f}{vmax:8.3f}{rng:9.3f}{delta:12.3f}{N:8d}")


Signal        Vmin    Vmax    Range    Δ (step)    Bits
-------------------------------------------------------
signal1      0.000  15.000   15.000       0.015      10
signal2      8.000  22.000   14.000       0.014      10
signal3      2.000 256.000  254.000       0.254      10


# Part 1: Results Summary

- **Signal 1**: Vmin = 0.000, Vmax = 15.000 → Range = 15.000 → 10 bits (Δ≈0.015)  
- **Signal 2**: Vmin = 8.000, Vmax = 22.000 → Range = 14.000 → 10 bits (Δ≈0.014)  
- **Signal 3**: Vmin = 2.000, Vmax = 256.000 → Range = 254.000 → 10 bits (Δ≈0.254)  

Using a 10-bit ADC for all three signals yields ≈1000 levels, giving a quantization step of about 0.1 % of each range and ensuring fine resolution without excessive data size.